File: src/lib/data/mock-responses.ts
import { GenerateResponse } from "@/lib/types";
import data from "./mock-responses.json";

const mockData = data.scenarios as Record<string, GenerateResponse>;
const defaultResponse = data.defaultResponse as GenerateResponse;

export function getMockResponse(
  patientId: string,
  goal: string,
  tone: string,
  channels: string[]
): GenerateResponse {
  // Try patient-specific mock with exact match first
  const specificKey = `${patientId}-${goal}-${tone}`;
  let response = mockData[specificKey];

  // Try same patient + same goal (any tone)
  if (!response) {
    const goalKey = Object.keys(mockData).find(
      (k) => k.startsWith(`${patientId}-${goal}-`)
    );
    if (goalKey) response = mockData[goalKey];
  }

  // Try same patient + same tone (any goal)
  if (!response) {
    const toneKey = Object.keys(mockData).find(
      (k) => k.startsWith(`${patientId}-`) && k.endsWith(`-${tone}`)
    );
    if (toneKey) response = mockData[toneKey];
  }

  // Fall back to patient's first available scenario
  if (!response) {
    const patientKeys = Object.keys(mockData).filter((k) =>
      k.startsWith(`${patientId}-`)
    );
    if (patientKeys.length > 0) {
      response = mockData[patientKeys[0]];
    }
  }

  // Final fallback to generic default
  if (!response) {
    response = defaultResponse;
  }

  // Filter to requested channels only
  return {
    ...response,
    generatedAt: new Date().toISOString(),
    channelMessages: response.channelMessages.filter((cm) =>
      channels.includes(cm.channel)
    ),
  };
}

File: backend/app/data/mock_responses.py
import json
from datetime import datetime, timezone
from pathlib import Path

from app.models import GenerateResponse, ChannelMessages, MessageVariant, Channel


_DATA_FILE = Path(__file__).parent / "mock_responses.json"
_data = json.loads(_DATA_FILE.read_text())
_scenarios: dict[str, dict] = _data["scenarios"]
_default_response: dict = _data["defaultResponse"]


def _parse_response(raw: dict, provider: str = "mock") -> GenerateResponse:
    """Parse a raw mock response dict into a GenerateResponse model."""
    channel_messages = [
        ChannelMessages(
            channel=cm["channel"],
            variants=[
                MessageVariant(
                    id=v["id"],
                    approach=v["approach"],
                    content=v["content"],
                    subject=v.get("subject"),
                    engagement_likelihood=v["engagementLikelihood"],
                    reasoning=v["reasoning"],
                )
                for v in cm["variants"]
            ],
        )
        for cm in raw["channelMessages"]
    ]
    return GenerateResponse(
        channel_messages=channel_messages,
        provider=provider,
        generated_at=datetime.now(timezone.utc).isoformat(),
    )


def get_mock_response(
    patient_id: str,
    goal: str,
    tone: str,
    channels: list[Channel],
) -> GenerateResponse:
    """Look up a mock response with smart fallback, mirroring the TypeScript implementation."""
    # Try exact match
    specific_key = f"{patient_id}-{goal}-{tone}"
    raw = _scenarios.get(specific_key)

    # Try same patient + same goal (any tone)
    if not raw:
        for key in _scenarios:
            if key.startswith(f"{patient_id}-{goal}-"):
                raw = _scenarios[key]
                break

    # Try same patient + same tone (any goal)
    if not raw:
        for key in _scenarios:
            if key.startswith(f"{patient_id}-") and key.endswith(f"-{tone}"):
                raw = _scenarios[key]
                break

    # Fall back to patient's first available scenario
    if not raw:
        for key in _scenarios:
            if key.startswith(f"{patient_id}-"):
                raw = _scenarios[key]
                break

    # Final fallback to generic default
    if not raw:
        raw = _default_response

    response = _parse_response(raw)

    # Filter to requested channels only
    response.channel_messages = [
        cm for cm in response.channel_messages if cm.channel in channels
    ]
    return response
